pkg/async/cond.go

```diff
@@ -0,0 +1,129 @@
+// Copyright 2023 Outreach Corporation. All Rights Reserved.
+
+// Description: Cond.go provides a context respecting sync condition
+
+package async
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+)
+
+// Cond mimics sync.Cond in purpose, with the added goals of easing usability slightly and respecting context expiry.
+//
+// It provides functionality similar sync.Cond
+//   - there is no signal method for waking a single Waiter
+//   - the Wait method exits with error if the context cancels.
+//   - it provides WaitForCondition, which intends to encapsulate the common pattern of acquiring a lock, checking
+//     a condition, and releasing the lock before waiting for a state change if the condition is not met.
+type Cond struct {
+	pointer atomic.Pointer[chan struct{}]
+	Mu      sync.Mutex
+}
+
+// ch returns the channel that Waiters are waiting on,
+// possibly creating one if it hasn't been initialized
+func (c *Cond) ch() chan struct{} {
+	// non atomic check for nil channel
+	load := c.pointer.Load()
+	if load == nil {
+		t := make(chan struct{})
+		c.pointer.CompareAndSwap(nil, &t)
+		return t
+	}
+	return *load
+}
+
+// Wait waits for the state change Broadcast until context ends.
+// If the returned error is non-nil, then the context ended before the state change Broadcast.
+func (c *Cond) Wait(ctx context.Context) error {
+	ch := c.ch()
+	select {
+	case <-ch:
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+// Broadcast signals the state change to all Waiters
+func (c *Cond) Broadcast() {
+	// swap in a new channel, close the old one
+	newChan := make(chan struct{})
+	ch := c.pointer.Swap(&newChan)
+	if ch != nil {
+		close(*ch)
+	}
+}
+
+// WaitForCondition acquires Cond's lock, then checks if the condition is true. If the condition is not true,
```

Comment:
Clarify what the lock actually is (and use the receiver name):
```suggestion
// WaitForCondition acquires c.Mu, then checks if the condition is true. If the condition is not true,
```